ddtrace/internal/rate_limiter.py

```diff
@@ -38,54 +38,57 @@ def __init__(self, rate_limit):
         self.tokens = rate_limit  # type: float
         self.max_tokens = rate_limit
 
-        self.last_update = compat.monotonic()
+        self.last_update_ns = compat.monotonic_ns()
 
-        self.current_window = 0  # type: float
+        self.current_window_ns = 0  # type: float
         self.tokens_allowed = 0
         self.tokens_total = 0
         self.prev_window_rate = None  # type: Optional[float]
 
         self._lock = threading.Lock()
 
-    def is_allowed(self):
-        # type: () -> bool
+    def is_allowed(self, timestamp_ns=None):
+        # type: (Optional[int]) -> bool
         """
         Check whether the current request is allowed or not
 
         This method will also reduce the number of available tokens by 1
 
+        :param int timestamp_ns: Optional timestamp in nanoseconds for the current request.
+            If not provided `compat.monotonic_ns()` is used.
         :returns: Whether the current request is allowed or not
         :rtype: :obj:`bool`
         """
+        if timestamp_ns is None:
+            timestamp_ns = compat.monotonic_ns()
         # Determine if it is allowed
-        allowed = self._is_allowed()
+        allowed = self._is_allowed(timestamp_ns)
         # Update counts used to determine effective rate
-        self._update_rate_counts(allowed)
+        self._update_rate_counts(allowed, timestamp_ns)
         return allowed
 
-    def _update_rate_counts(self, allowed):
-        # type: (bool) -> None
-        now = compat.monotonic()
-
+    def _update_rate_counts(self, allowed, timestamp_ns):
+        # type: (bool, int) -> None
         # No tokens have been seen yet, start a new window
-        if not self.current_window:
-            self.current_window = now
+        if not self.current_window_ns:
+            self.current_window_ns = timestamp_ns
```

Comment:
Hmm this could really be an issue if a custom span start is used. I don't think it's very common to do it but we have to consider this since it can.

I suppose in the adversary case here a user could specify a span time well into the future and then create a bunch well into the past. The future one would move the window forward and reset the tokens. The past ones would all then be permitted.

I don't think there's much we can/should do to address it other than yell through logs if it breaks something in the limiting.

Reply:
it probably isn't a problem, unless we think the non-monotonic clock is going to cause issues.

since we only call `sampler.sample(span)` in `Tracer.start_span` right after we have created a new span, and there is no chance that someone has been able to adjust the span start time.

Reply:
Ah right, the start time can be passed to the initializer of span but we don't expose it through `trace`/`start_span` :+1:

Hmm I really have no idea how sensitive clocks are in practice - surely at scale it'll occur but I don't think the impacts will be noticeable or frequent.